src/hotspot/share/utilities/debug.cpp

```diff
@@ -296,20 +296,49 @@ void report_java_out_of_memory(const char* message) {
 
 class Command : public StackObj {
  private:
-  ResourceMark _rm;
+  static int level;
   DebuggingContext _debugging;
+  bool _has_rm;
+  // ResourceMark is only created if a Thread or a JavaThread is required,
+  // and we are actually on a Thread.
+  union { ResourceMark _rm; };
  public:
-  static int level;
-  Command(const char* str) {
-    if (level++ > 0)  return;
-    tty->cr();
-    tty->print_cr("\"Executing %s\"", str);
+  Command(const char* str) : _has_rm(false) {
+    if (level++ == 0) {
+      tty->cr();
+      tty->print_cr("\"Executing %s\"", str);
+    }
+    tty->flush();
   }
-
   ~Command() {
+    if (_has_rm) _rm.~ResourceMark();
     tty->flush();
     level--;
   }
+
+  bool onThread() {
+    Thread* thread = Thread::current_or_null();
+    if (thread == nullptr) {
+      tty->print_cr("Failed: Current thread is not attached");
+      return false;
+    }
+
+    if (!_has_rm) {
+      ::new (&_rm) ResourceMark();
+      _has_rm = true;
+    }
+    return true;
+  }
+
+  bool onJavaThread() {
+    if (!onThread()) return false;
+
+    if (JavaThread::active() == nullptr) {
+      tty->print_cr("Failed: Current thread is not a java thread or the vm thread");
+      return false;
+    }
+    return true;
+  }
```

Comment:
I don't think we need this. The commands that require a JavaThread should be checking that directly themselves. Typically we assume/expect the person debugging to know what they are dealing with and use the appropriate commands.

Reply:
I will remove this function and do the check in the 3 call sites of it.

Reply:
Thank you for the clarification! That makes sense.

I'll move forward with David's suggestion to have the duplication where the `Command` objects are used.